engine/crates/runtime/src/user_hooks.rs

```diff
@@ -51,4 +51,10 @@ pub trait UserHooksImpl: Send + Sync {
     type Context;
 
     async fn on_gateway_request(&self, headers: HeaderMap) -> Result<(Self::Context, HeaderMap), UserHookError>;
+
+    async fn on_authorization(
```

Comment:
IMHO the function should be:
```
async fn authorized&self, context: &mut Self::Context, input: impl ExactSizeIterator<Item = impl serde::Serialize>) -> ...
```
Might as well use the directive name, engine will provide a &mut Self::Context. The need to use `std::mem::take` is an implementation details of wasm IMHO. No need either to specify the types of of input either. How it's serialized is also up to the implementation IMO.

Why differentiate `UserError` from `UserHookError`? In all cases it'll end up as a GraphQL error 🤷

Reply:
IMHO would just `UserError` as `HookError` and that's it.

Reply:
The `UserError` gives user control to define the message and the extensions.

Reply:
Well yes, but in the end it doesn't make a difference to the engine whether it's wasm error or a hook error. It'll be treated the same and transformed into a GraphqlError with message & extensions. so as it'll be done in all cases we could simplify the types here IMHO. 🤷